File: C/palindrome_linked_list.c
#include <stdbool.h>
#include <stddef.h>


struct ListNode {
    int val;
    struct ListNode *next;
};

bool isPalindrome(struct ListNode* head) {
    struct ListNode *slow = head;
    struct ListNode *fast = head;
    while (fast != NULL && fast->next != NULL) {
        slow = slow->next;
        fast = fast->next->next;
    }
    struct ListNode *cur = slow;
    struct ListNode *prev = NULL;
    while (cur != NULL) {
        struct ListNode *temp = cur->next;
        cur->next = prev;
        prev = cur;
        cur = temp;
    }
    struct ListNode *l = head;
    struct ListNode *r = prev;
    while (r != NULL) {
        if (l->val != r->val) {
            return false;
        }
        l = l->next;
        r = r->next;
    }
    return true;
};

File: JavaScript/palindrome_linked_list.js
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

/**
 * @param {ListNode} head
 * @return {boolean}
 */

var isPalindrome = function (head) {
    let slow = head
    let fast = head
    while (fast !== null && fast.next !== null) {
        slow = slow.next
        fast = fast.next.next
    }

    let prev = null
    let cur = slow
    while (cur !== null) {
        let temp = cur.next
        cur.next = prev
        prev = cur
        cur = temp
    }

    let left = head
    let right = prev

    while (left !== null && right !== null) {
        if (left.val !== right.val) {
            return false
        }
        left = left.next
        right = right.next
    }
    return true
};